examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/DeviceSimulatorRepository.java

```diff
@@ -78,12 +79,17 @@ public AbstractSimulatedDevice remove(final String tenant, final String id) {
         return devices.remove(new DeviceKey(tenant.toLowerCase(), id));
     }
 
+    public Set<String> getTenants() {
+        return tenants;
```

Comment:
instead storing the tenant to an extra set we could retrieve the tenants on the fly from the devices map.
```
public Set<String> getTenants() {
        devices.keySet().stream().map(key -> key.tenant).collect(Collectors.toSet());
    }
```

Reply:
Sounds pretty expensive. I for instance simulate 500k devices most of the time.